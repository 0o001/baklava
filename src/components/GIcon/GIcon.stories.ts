import { storiesOf } from '@storybook/vue';
import GIcon from './GIcon.vue';

storiesOf('GIcon', module)
  .addParameters({ component: GIcon })
  .add('Example', () => ({
    components: { GIcon },
    template: `<div> <GIcon name="alert-circle" strokeWidth='2px' /> <GIcon stroke="black" size="44px" fill="var(--orange-100)" name="star" /> </div>`,
  }))