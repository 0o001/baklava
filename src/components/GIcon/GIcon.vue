<template>
  <svg
    v-bind="this.attrs"
    v-on="$listeners"
    :class="classNames"
    :dataName="this.name"
    :dataTags="tags"
    :dataType="this.name"
    v-html="icon.contents"
  />
</template>

<script lang="ts">

import feather from 'feather-icons';

export default {
  name: 'GIcon',
  props: {
    animation: {
      type: String,
      default: undefined,
    },
    animationSpeed: {
      type: String,
      default: undefined,
    },
    fill: {
      type: String,
      default: 'none',
    },
    size: {
      type: [Number, String],
      default: '24px',
    },
    stroke: {
      type: String,
      default: 'currentColor',
    },
    strokeLinecap: {
      type: String,
      default: 'round',
    },
    strokeLinejoin: {
      type: String,
      default: 'round',
    },
    strokeWidth: {
      type: [Number, String],
      default: 2,
    },
    tag: {
      type: String,
      default: 'i',
    },
    name: {
      type: String,
      required: true,
      validator (name) {
        if (!feather.icons[name]) {
          throw new Error(`"${name}" is not an available icon type.`);
        }
        return true;
      },
    },
    color: {
      type: String,
      default: undefined,
    },
  },
  computed: {
    icon () {
      return feather.icons[this.name];
    },
    tags () {
      return feather.icons[this.name] ? feather.icons[this.name].tags : '';
    },
    classNames () {
      const classes = {
        'g-icon': true,
        feather: true,
        [`feather--${this.name}`]: true,
        [`feather--${this.animation}`]: this.animation,
        [`feather--${this.animationSpeed}`]: this.animationSpeed,
      };
      return classes;
    },
    attrs () {
      return {
        ...this.icon.attrs,
        fill: this.fill,
        height: this.size,
        stroke: this.color ? `var(--${this.color})` : this.stroke,
        'stroke-linecap': this.strokeLinecap,
        'stroke-linejoin': this.strokeLinejoin,
        'stroke-width': this.strokeWidth,
        width: this.size,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@keyframes feather--spin {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
.feather {
  display: inline-block;
  overflow: hidden;
  &--spin {
    animation: feather--spin 2s linear infinite;
  }
  &--pulse {
    animation: feather--spin 2s infinite steps(8);
  }
  &--slow {
    animation-duration: 3s;
  }
  &--fast {
    animation-duration: 1s;
  }
  &__content {
    display: block;
    height: inherit;
    width: inherit;
  }
}
</style>
