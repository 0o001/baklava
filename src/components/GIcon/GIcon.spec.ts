import { shallowMount } from '@vue/test-utils';
import GIcon from './';

describe('GIcon', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(GIcon, {
      propsData: {
        name: 'alert-circle'
      }
    });
  });

  it('should match snapshot when show is prop is given as false ', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render base color as default when color prop does not set', () => {
    expect(wrapper.vm.attrs.stroke).toEqual('currentColor');
  });

  it('should render given color when color prop set', async () => {
    // given
    wrapper.setProps({
      color: 'green-500',
    });

    // when
    await wrapper.vm.$nextTick();

    // then
    expect(wrapper.vm.attrs.stroke).toEqual('var(--green-500)');
  });
});
