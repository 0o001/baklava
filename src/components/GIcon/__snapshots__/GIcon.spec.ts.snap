// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIcon should match snapshot when show is prop is given as false  1`] = `
<svg
  class="g-icon feather feather--alert-circle"
  dataName="alert-circle"
  dataTags="warning,alert,danger"
  dataType="alert-circle"
  fill="none"
  height="24px"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24px"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="12"
    cy="12"
    r="10"
  />
  <line
    x1="12"
    x2="12"
    y1="8"
    y2="12"
  />
  <line
    x1="12"
    x2="12.01"
    y1="16"
    y2="16"
  />
</svg>
`;
